File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/Mission/BaseMission.java
package jp.aoichaan0513.A_TosoGame_Live.Mission;

import jp.aoichaan0513.A_TosoGame_Live.Main;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

public class BaseMission {

    public MissionTimer runner = null;
    private static BukkitTask timer = null;

    public boolean isStarted = false;

    public void start() {
        if (timer == null) {
            runner = new MissionTimer(0);
            timer = runner.runTaskTimer(Main.getInstance(), 0L, 20L);
            return;
        }
        return;
    }

    public void end() {
        if (timer != null) {
            timer.cancel();
            timer = null;
            runner = null;
            return;
        }
        return;
    }

    public boolean isStarted() {
        return isStarted;
    }

    public void setStarted(boolean started) {
        isStarted = started;
    }

    public static class MissionTimer extends BukkitRunnable {

        private static int initialTime;
        private static int time;

        public MissionTimer(int time) {
            this.initialTime = time;
            this.time = time;
        }

        @Override
        public void run() {
            time--;
        }
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/Mission/BaseMission.kt
package jp.aoichaan0513.A_TosoGame_Live.Mission

import jp.aoichaan0513.A_TosoGame_Live.Main
import org.bukkit.scheduler.BukkitRunnable
import org.bukkit.scheduler.BukkitTask

class BaseMission {
    companion object {

        private var timer: BukkitTask? = null
        var isStart = false
            private set

        fun startMission() {
            if (timer != null) return
            timer = BaseMissionRunnable((20 * 60) * 0).runTaskTimer(Main.pluginInstance, 0L, 20L)
            isStart = true
        }

        fun endMission() {
            if (timer == null) return
            timer!!.cancel()
            timer = null
        }

        fun resetMission() {
            if (timer != null) return
            isStart = false
        }

        fun endAndResetMission() {
            HunterZone.endMission()
            HunterZone.resetMission()
        }


        private class BaseMissionRunnable(private val initialMissionTime: Int) : BukkitRunnable() {
            private var missionTime: Int

            init {
                missionTime = initialMissionTime
            }

            override fun run() {
                missionTime--
            }
        }
    }
}